packages/plugin/vite/src/VitePlugin.ts

```diff
@@ -0,0 +1,215 @@
+import { default as vite } from 'vite';
+
+import { VitePluginConfig } from './Config';
+import ViteConfigGenerator from './ViteConfig';
+
+const d = debug('electron-forge:plugin:vite');
+const DEFAULT_PORT = 3000;
+
+export default class VitePlugin extends PluginBase<VitePluginConfig> {
+  name = 'vite';
+
+  private isProd = false;
+
+  // The root of the Electron app
+  private projectDir!: string;
+
+  // Where the Vite output is generated. Usually `${projectDir}/.vite`
+  private baseDir!: string;
+
+  private configGenerator: Promise<ViteConfigGenerator>;
+
+  private watchers: RollupWatcher[] = [];
+
+  private servers: http.Server[] = [];
+
+  constructor(c: VitePluginConfig) {
+    super(c);
+
+    this.configGenerator = new Promise((resolve) => {
+      // eslint-disable-next-line promise/catch-or-return
+      vite.resolveConfig({ configFile: this.config.renderer.config }, 'serve').then((resolvedConfig) => {
+        // Get the user to set port in vite config file.
+        resolve(new ViteConfigGenerator(this.config, this.projectDir, this.isProd, resolvedConfig.server.port ?? DEFAULT_PORT));
+        return;
+      });
+    });
+  }
+
+  init = (dir: string): void => {
+    this.setDirectories(dir);
+
+    d('hooking process events');
+    process.on('exit', (_code) => this.exitHandler({ cleanup: true }));
+    process.on('SIGINT' as NodeJS.Signals, (_signal) => this.exitHandler({ exit: true }));
+  };
+
+  setDirectories = (dir: string): void => {
+    this.projectDir = dir;
+    this.baseDir = path.resolve(dir, '.vite');
+  };
+
+  getHooks = (): ForgeMultiHookMap => {
+    return {
+      prePackage: [
+        namedHookWithTaskFn<'prePackage'>(async () => {
+          this.isProd = true;
+          fs.rmSync(this.baseDir, { recursive: true, force: true });
+
+          await this.compileMain();
+          await this.compileRenderers();
+        }, 'Building vite bundles'),
+      ],
+    };
+  };
+
+  private alreadyStarted = false;
+  startLogic = async (): Promise<StartResult> => {
+    if (this.alreadyStarted) return false;
+    this.alreadyStarted = true;
+
+    fs.rmSync(this.baseDir, { recursive: true, force: true });
+
+    return {
+      tasks: [
+        {
+          title: 'Compiling main process code',
+          task: async () => {
+            await this.compileMain(true);
+          },
+          options: {
+            showTimer: true,
+          },
+        },
+        {
+          title: 'Launching dev servers for renderer process code',
+          task: async () => {
+            await this.launchRendererDevServers();
+          },
+          options: {
+            persistentOutput: true,
+            showTimer: true,
+          },
+        },
+      ],
+      result: false,
+    };
+  };
+
+  compileMain = async (watch = false): Promise<void> => {
+    const buildResult = await vite.build({
+      configFile: this.config.mainConfig,
+
+      // However, bllow config options aways merge into `configFile`.
```

Comment:
Haha, you found the note. This is for reviewser to see(you can remove it), the configuration written in the code will have a higher priority than the user's `configFile`, which means it may be opinionated.

Reply:
I think that's fine. We do this in a lot of places within Forge.